File: 백준/Silver/2003. 수들의 합 2/수들의 합 2.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N=Integer.parseInt(st.nextToken());
        long M=Long.parseLong(st.nextToken());

        long[] arr=new long[N];
        st=new StringTokenizer(br.readLine());
        for(int i=0; i<N; i++) arr[i]=Long.parseLong(st.nextToken());

        int count=0;
        int start=0;
        int end=0;
        long sum=0;
        while(true){
            if(sum<M) {
                if(end==N) break;
                sum+=arr[end];
                end++;
            } else if (sum>M){
                sum-=arr[start];
                start++;
            } else {
//                System.out.println("이때: "+start+" "+(end-1));
                count++;
                sum-=arr[start];
                start++;
            }
        }
        System.out.println(count);
        br.close();
    }
}

File: 백준/Silver/2003. 수들의 합 2/수들의 합 2.cc
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main()
{
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	int N, M;
	cin >> N >> M;

	vector<int> arr;
	for (int i = 0; i < N; i++) {
		int next;
		cin >> next;
		arr.push_back(next);
	}

	int count = 0;
	int start = 0;
	int end = 0;
	long long sum = arr[start];
	while (end < N) {
		if (sum < M) {
			/*cout << "sum < M" << "\n";
			cout << "start: " << start << ", end: " << end << ", sum: " << sum << "\n";*/
			end++;
			if (end == N) break;
			sum += arr[end];
		}
		else if (sum > M) {
			/*cout << "sum > M" << "\n";
			cout << "start: " << start << ", end: " << end << ", sum: " << sum << "\n";*/
			if (start == end) {
				end++;
				if (end == N) break;
				sum += arr[end];
			}
			else {
				sum -= arr[start];
				start++;
			}
		}
		else {
			/*cout << "sum == M" << "\n";
			cout << "start: " << start << ", end: " << end << ", sum: " << sum << "\n";*/
			count++;
				end++;
				if (end == N) break;
				sum += arr[end];

			
		}
	}

	cout << count << "\n";
	
}
